Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h

```diff
@@ -633,131 +635,206 @@ clip(TriangleMesh& tm,
+                                          .concurrency_tag(Concurrency_tag()));
+
+
+  if (allow_self_intersections)
+    clip_volume=false;
+
+  if (clip_volume && !is_closed(pm)) clip_volume=false;
+  if (clip_volume && !use_compact_clipper) use_compact_clipper=true;
+
+
+  auto fcc = get(dynamic_face_property_t<std::size_t>(), pm);
+
+  std::size_t nbcc = connected_components(pm, fcc, CGAL::parameters::edge_is_constrained_map(ecm));
+
+  std::vector<bool> classified(nbcc, false);
+  std::vector<std::size_t> ccs_to_remove;
+
+  for (auto f : faces(pm))
   {
-    case ON_NEGATIVE_SIDE:
-      return true; // nothing to clip, the full mesh is on the negative side
-    case ON_POSITIVE_SIDE:
-      remove_all_elements(tm); // clear the mesh that is fully on the positive side
-      return true;
-    default:
-      break;
+    std::size_t ccid = get(fcc, f);
+    if (classified[ccid]) continue;
+    halfedge_descriptor hf = halfedge(f, pm);
+    for(halfedge_descriptor h : CGAL::halfedges_around_face(hf, pm))
+    {
+      CGAL::Oriented_side os = get(vos, target(h,pm));
+      if (os==CGAL::ON_ORIENTED_BOUNDARY) continue;
+      classified[ccid]=true;
+      if (os==CGAL::ON_POSITIVE_SIDE) ccs_to_remove.push_back(ccid);
+    }
+
+    if (!classified[ccid])
+    {
+      if (!use_compact_clipper) ccs_to_remove.push_back(ccid);
+      classified[ccid]=true;
+    }
   }
 
-  const bool do_not_modify = choose_parameter(get_parameter(np, internal_np::allow_self_intersections), false);
-  return clip(tm, clipper, np, params::do_not_modify(do_not_modify));
+  remove_connected_components(pm, ccs_to_remove, fcc);
+
+  if (clip_volume)
+  {
+    std::vector<halfedge_descriptor> borders;
+    extract_boundary_cycles(pm, std::back_inserter(borders));
+
+    for (halfedge_descriptor h : borders)
+    {
+      Euler::fill_hole(h, pm);
+#ifndef CGAL_PLANE_CLIP_DO_NOT_USE_TRIANGULATION
+      if (triangulate)
+        triangulate_face(face(h,pm), pm, parameters::vertex_point_map(vpm).geom_traits(traits));
+#endif
+    }
+  }
+
+  return true;
 }
 
 /**
   * \ingroup PMP_corefinement_grp
   *
   * \brief clips `tm` by keeping the part that is inside `iso_cuboid`.
   *
-  * If `tm` is closed, the clipped part can be closed too if the named parameter `clip_volume` is set to `true`.
+  * If `tm` is closed, the clipped part can be kept closed by setting the named parameter `clip_volume` to `true`.
   * See Subsection \ref coref_clip for more details.
   *
-  * \note `Iso_cuboid_3` must be from the same %Kernel as the point of the internal vertex point map of `TriangleMesh`.
-  * \note `Iso_cuboid_3` must be from the same %Kernel as the point of the vertex point map of `tm`.
+  * \note `Iso_cuboid_3` must be from the same kernel as the point of the internal vertex point map of `TriangleMesh`.
+  * \note `Iso_cuboid_3` must be from the same kernel as the point of the vertex point map of `tm`.
```

Comment:
you removed that in the other functions

Reply:
it is needed as internally make_hexahedron is used